external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala

```diff
@@ -88,7 +92,19 @@ private[kafka010] abstract class KafkaRowWriter(
       throw new NullPointerException(s"null topic present in the data. Use the " +
         s"${KafkaSourceProvider.TOPIC_OPTION_KEY} option for setting a default topic.")
     }
-    val record = new ProducerRecord[Array[Byte], Array[Byte]](topic.toString, key, value)
+    val record = if (projectedRow.isNullAt(3)) {
+      new ProducerRecord[Array[Byte], Array[Byte]](topic.toString, null, key, value)
+    } else {
+      val headerArray = projectedRow.getArray(3)
+      val headers = (0 until headerArray.numElements()).map(
+        i => {
+          val struct = headerArray.getStruct(i, 2)
+          new RecordHeader(struct.getUTF8String(0).toString, struct.getBinary(1))
+            .asInstanceOf[Header]
```

Comment:
Do you need this cast? it's already a `Header`? I could be missing something

Reply:
@srowen Yes, for the limitation of Java type system, `Iterable[RecordHeader]` is not regarded as `Iterable[Header]`. It's why.